ui/src/ducks/app/salt.js

```diff
@@ -1,31 +1,267 @@
+
 // Actions
-const ADD_JOB = 'REMOVE_JOB';
-const REMOVE_JOB = 'UPDATE_JOBS';
+export const ADD_JOB = 'ADD_JOB';
+export const REMOVE_JOB = 'REMOVE_JOB';
+const ADD_JOB_EVENT = 'ADD_JOB_EVENT';
+const CONNECT_SALT_API = 'CONNECT_SALT_API';
+const SET_JOB_STATUS = 'SET_JOB_STATUS';
+export const JOB_COMPLETED = 'JOB_COMPLETED';
 
 // Reducer
 const defaultState = {
-  jobs: []
+  jobs: [],
+  connected: false,
 };
 
 export default function reducer(state = defaultState, action = {}) {
   switch (action.type) {
     case ADD_JOB:
-      return { ...state, jobs: [...state.jobs, action.payload] };
+      return {
+        ...state,
+        jobs: [...state.jobs, action.payload],
+      };
     case REMOVE_JOB:
       return {
         ...state,
-        jobs: state.jobs.filter(job => job !== action.payload)
+        jobs: state.jobs.filter(job => job.jid !== action.payload.jid),
       };
+    case ADD_JOB_EVENT:
+      return {
+        ...state,
+        jobs: state.jobs.map(job =>
+          job.jid === action.payload.jid
+            ? { ...job, events: [...job.events, action.payload.event] }
+            : job,
+        ),
+      };
+    case JOB_COMPLETED:
+      return {
+        ...state,
+        jobs: state.jobs.map(job =>
+          job.jid === action.payload.jid
+            ? {
+                ...job,
+                completed: true,
+                completedAt: action.payload.completedAt,
+                status: action.payload.status,
+              }
+            : job,
+        ),
+      };
+    case SET_JOB_STATUS:
+      return {
+        ...state,
+        jobs: state.jobs.map(job =>
+          job.jid === action.payload.jid
+            ? { ...job, status: action.payload.status }
+            : job,
+        ),
+      };
+    case CONNECT_SALT_API:
+      return { ...state, connected: true };
     default:
       return state;
   }
 }
 
 // Action Creators
+const defaultJob = {
+  type: '',
+  jid: null,
+  completed: false,
+  completedAt: null,
+  status: {},
+  events: [],
+};
+
 export function addJobAction(job) {
-  return { type: ADD_JOB, payload: job };
+  const payload = { ...defaultJob, ...job };
+  if (payload.completedAt !== null) {
+    payload.completed = true;
+  }
+  return { type: ADD_JOB, payload };
 }
 
 export function removeJobAction(job) {
   return { type: REMOVE_JOB, payload: job };
 }
+
+export function connectSaltApiAction(payload) {
+  return { type: CONNECT_SALT_API, payload };
+}
+
+function addJobEventAction(jid, event) {
+  return { type: ADD_JOB_EVENT, payload: { jid, event } };
+}
+
+export function setJobCompletedAction(jid, completedAt, status) {
+  return { type: JOB_COMPLETED, payload: { jid, completedAt, status } };
+}
+
+export function setJobStatusAction(jid, status) {
+  return { type: SET_JOB_STATUS, payload: { jid, status } };
+}
+
+// Selectors
+export const allJobsSelector = state => state.app.salt.jobs;
+
+// Sagas
+export function* initialize() {
```

Comment:
you need to get a `payload` here and pass it to `yield fork(watchSaltEvents, payload);`